lib/ringbuf/src/stringbuf.rs

```diff
@@ -0,0 +1,159 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+pub use core::fmt;
+pub use core::fmt::Write;
+/// Re-export the bits we use so that code generated by the
+/// macros is guaranteed to be able to find them.
+pub use userlib::util::StaticCell;
+
+/// Declares a ringbuffer in the current module or context.
+///
+/// `stringbuf!(NAME, N, expr)` makes a ringbuffer named `NAME`,
+/// with room for `N` such entries, all of which are initialized to `expr`.
+///
+/// The resulting ringbuffer will be static, so `NAME` should be uppercase. If
+/// you want your ringbuffer to be detected by Humility's automatic scan, its
+/// name should end in `STRINGBUF`.
+///
+/// The actual type of `name` will be `StaticCell<Stringbuf<N>>`.
+///
+/// To support the common case of having one quickly-installed ringbuffer per
+/// module, if you omit the name, it will default to `LOG__STRINGBUF`.
+#[cfg(not(feature = "disabled"))]
+#[macro_export]
+macro_rules! stringbuf {
+    ($name:ident, $n:expr, $init:expr) => {
+        #[used]
+        static $name: $crate::StaticCell<$crate::stringbuf::Stringbuf<$n>> =
+            $crate::StaticCell::new($crate::stringbuf::Stringbuf {
+                last: None,
+                buffer: [0; $n],
+            });
+    };
+    ($n:expr, $init:expr) => {
+        $crate::stringbuf!(LOG__STRINGBUF, $n, $init);
+    };
+}
+
+#[cfg(not(feature = "disabled"))]
+#[macro_export]
+macro_rules! stringbuf_root {
+    ($name:ident, $n:expr, $init:expr) => {
+        #[used]
+        pub static $name: $crate::StaticCell<$crate::stringbuf::Stringbuf<$n>> =
+            $crate::StaticCell::new($crate::stringbuf::Stringbuf {
+                last: None,
+                buffer: [0; $n],
+            });
+    };
+    ($n:expr, $init:expr) => {
+        $crate::stringbuf!(LOG__STRINGBUF, $n, $init);
+    };
+}
+
+#[cfg(feature = "disabled")]
+#[macro_export]
+macro_rules! stringbuf {
+    ($name:ident, $n:expr, $init:expr) => {
+        #[allow(dead_code)]
+        const _: $t = $init;
```

Comment:
What is `$t` supposed to be? Wouldn't this lead to a build failure?

Reply:
Yep, should have tried to build with `disabled` feature.